include/bitcoin/system/impl/stream/streamers/byte_reader.ipp

```diff
@@ -179,7 +179,9 @@ template <typename IStream>
 template <size_t Size>
 data_array<Size> byte_reader<IStream>::read_reverse() noexcept
 {
-    return std::move(system::reverse(read_forward<Size>()));
+    auto data = read_forward<Size>();
+    auto& reverse = system::reverse(data);
+    return std::move(reverse);
```

Comment:
This changes reverse to the copy overload, which should be avoided.

The last two lines should be collapsible with a simple return, as copy elision is assumed. It shouldn’t have been necessary in the original. Is it possible that was the issue?

Reply:
It is my understanding that this is the equivalent to what was provided while avoiding non-const lvalues which aren't supported by GCC under any condition that I'm aware of.  I've opened an issue for tracking within libbitcoin-build relating to escalating the documented warning in Visual Studio to an error for the purposes of maintaining cross platform compatibility.

Reply:
Oh, this is not move overloaded. The problem is simply that the move needs to be removed. The return is a reference.

Reply:
Oh, but the origin (read) is an r-value, so the problem is that reverse doesn’t have a move overload, which would return an r-value (and therefore would not be moved out, relying on local variable copy elision.

Reply:
Sorry, working from my phone. These three functions (distinct, reverse, sort) were originally move overloaded, and I was assuming from the test cases this was still the case. I must have changed them to const/non-const refs only, which silently broke these tests on msvc.

Reply:
Ideally the non-copy variants would each have a move overload (as well as the two existing fns). This would have the same non-copy efficiency as the existing non-const fn, while allowing r-values to be passsed. I think that addition move overloads for all three is the proper resolution. I also noticed that std lib provides a std::reverse_copy, which should be used.